Assignment5

DaisyPalencia opened the pull request:
Pairing with @AnnikaStachowiak

jacobnewcomb reviewed (commented):
- Create a class library project called "GenericsHomework.". ✔
- Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.✔
- Define the Node class✔
- The values should be homogeneous.✔
- Include a constructor that takes a value. (No validation is necessary on the value). ✔
- Add a ToString() override that writes out the value's ToString() result. ✔
- Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
- The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
- The Next property setter should be private. ✔
- Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
- Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
- Whether it is sufficient to only set Next to itself ✔
- Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
- Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
- Create an Exists method to test to see if a value exists in the list. ✔
- Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
- You should not rely on any BCL generic classes for your implementation. ✔

Everything looks great! I really like how you did your clear method, I think I will implement this strategy in my own fork!

alowe4 commented on GenericsHomework/GenericsHomework.Tests/GenericsHomeworkTests.csproj:
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
why are there two test csproj files????

DaisyPalencia replied:
I think this is just a mistake? It doesn't show up when I look for it.

alowe4 commented on GenericsHomework/GenericsHomework.Tests/GenericsHomework.Tests.csproj:
```diff
@@ -0,0 +1,24 @@
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
```
needs EnableNETAnalyzers, and LangVersion 11

DaisyPalencia replied:
thank you, I missed that!

alowe4 reviewed (commented):
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.✔
 Define the Node class✔
 The values should be homogeneous.✔	
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
 The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
 The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:✔
 Whether it is sufficient to only set Next to itself ✔
 Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
 Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ❌[no comment exists for this]
Create an Exists method to test to see if a value exists in the list. ✔[you made one, it just doesn't work]
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔


Fundamentals

Place all shared project properties into a Directory.Build.Props file.❌ 
Place all shared project items into a Directory.Build.targets file.✔
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

---------------------------------------------------------------------------
main things to fix 
 -the lack of a Directory.Build.Props file,
 -the 2 test csproj files
 -the lack of testing on append and exists
 -missing comment about garbage collection

DaisyPalencia commented on GenericsHomework/GenericsHomework/Node.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        public T Value { get; }
+
+        public Node<T> Next { get; private set; }
```
We wanted to have it use the "?" so the code would be more understandable and clearer.

DaisyPalencia commented on GenericsHomework/GenericsHomework.Tests/NodeTests.cs:
```diff
@@ -0,0 +1,53 @@
+    }
+    /*
+    [TestMethod]
+    public void NodeAppendPass()
```
yeah we couldn't get that test to work before Monday, so it was commented out. Thank you, we will go back and fix that!

DaisyPalencia replied:
tests added for exists and appends and clear

DaisyPalencia commented on GenericsHomework/GenericsHomework/Node.cs:
```diff
@@ -0,0 +1,66 @@
+                {
+                    return true;
+                }
+                else if (currentNode.Next != null)
```
okay, fixed!

DaisyPalencia commented:
> main things to fix 
-the lack of a Directory.Build.Props file, 
-the 2 test csproj files -the lack of testing on append and exists 
-missing comment about garbage collection

Added the comment for garbage collection, added Directory.Build.Props file, removed second .csproj. 

Thank you for your comments they were very helpful and made our code a lot clearer and cleaner.

mmwoodfo commented on GenericsHomework/GenericsHomework/Node.cs:
```diff
@@ -0,0 +1,85 @@
+{
+    public T Value { get { return valueActual; } }
+
+    public Node<T> next;
```
```suggestion
    private Node<T> _Next;
```

mmwoodfo commented on GenericsHomework/GenericsHomework/Node.cs:
```diff
@@ -0,0 +1,85 @@
+        get { return next; }
+        private set { next = value; }
+    }
+    public override string ToString()
+    {
+        return valueActual?.ToString() ?? "is null";
+    }
```
```suggestion
    public override string? ToString()
    {
        return valueActual?.ToString();
    }
```
